### PR TITLE
Explore: Refactor & centralize URL/state sync

### PR DESCRIPTION
**Pull request by Elfo404**
**What is this feature?**

Overhauls Explore's URL & state sync by adding a redux middleware and centralizing the logic in `useStateSync`.
This will be needed for a follow up PR that will introduce URL versioning & migrations for Explore.

While the hook's logic is a bit complex (it may be refactored a bit and made simpler), having the sync logic centralized makes it easier to maintain.

notable differences:
- remove compact URLs warning messages (and related store properties) as we decided to not remove them but handle them in migrations (to be implemented in a follow up PR)
- history back and forward are way more consistent, new entries in history only pushed when needed. (try closing a split pane and press the back button, then forward button)

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/ExplorePage.test.tsx
Comment (gelicia) [inline public/app/features/explore/ExplorePage.test.tsx:118]: 😭

File: public/app/features/explore/ExplorePage.tsx
Comment (gelicia) [inline public/app/features/explore/ExplorePage.tsx:76]: ✨ foreshadowing ✨

**Comment by gelicia** [comment]
Is this expected to fix the bug with slow queries and closing panes? I'm still seeing that. 

I looked over the code. It looks good, and I can do more extensive testing of it tomorrow.

**Comment by Elfo404** [comment]
@gelicia no, that would require decoupling the exploreId from its rendering, something we'd need migrations first. this should have no real behavior changes aside from centralizing the logic and those minor fixes mentioned in the description

**Comment by gelicia** [comment]
It used to be that clicking explore on the side twice would reset the datasource to the default one, I'm not seeing that behavior anymore. 

Old behavior (using play)
1. Go to explore and select a datasource. 
2. Click the top icon to go home
3. Open the toolbar and select explore. It shows the datasource selected in 1
4. Open the toolbar and select explore again. It shows the default datasource.

This branch seems to always show the last selected datasource.

I was also intermittently getting into a situation where I'd close out of split view and it would reopen, with the right pane being the empty explore page. I will edit this if I can reliably replicate it.

**Comment by Elfo404** [comment]
interesting finding. i don't think it was intentional though, we always wanted the last used datasource to come up when opening explore, not the default one. when i was writing this i intentionally made sure that clicking on explore in the navbar would open a "clean" explore pane with the last selected datasource 🤔 do you think it is a blocker?

**Comment by Elfo404** [comment]
> I was also intermittently getting into a situation where I'd close out of split view and it would reopen, with the right pane being the empty explore page. I will edit this if I can reliably replicate it.

i think this is more or less the same bug we have in main when closing a right pane with a pending query, the difference being that i main it shows a toolbar in a wrong state and allows to close also the remaining pane, forcing you to either refresh the page or exit and reopening explore

https://github.com/grafana/grafana/assets/1170767/e0b9c5d3-5321-4cba-8096-31bf6ef46fd9

while here it shows a right pane with the empty ds list.

it was one of the first thing i worked on, i think something went missing during a rebase

**Comment by Elfo404** [comment]
found it and fixed it on [44d04da](https://github.com/grafana/grafana/pull/66286/commits/44d04da39d28c1de5d02dcacfd435f83f4644267)

although it doesn't really fix the root cause of the already known bugs (that needs migrations to be properly fixed)


in this example, closing the left pane makes it so that the right pane becomes left, but the left pane's query is still running, and when it finishes it updates the left panes (while it shouldn't)
https://github.com/grafana/grafana/assets/1170767/4b515b8b-47bd-43bf-bb5a-3908d46b9cff

**Review by gelicia — APPROVED** [review]
good work! We will wait feedback if anyone else was using "click explore twice to use defualt datasource" or if it was just me 😅

**Review by mckn — APPROVED** [review]
Wow, awesome work with this! I have started to clean up https://github.com/grafana/grafana/pull/65524 but will wait until this PR is merged.

File: packages/grafana-data/src/types/explore.ts
Comment (ifrost) [inline packages/grafana-data/src/types/explore.ts:11]: Where do we use `DataSourceRef` instead of a string? Can't seem to find it 🤔

File: public/app/features/explore/hooks/useStopQueries.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStopQueries.ts:10]:
```suggestion
 * This avoids unnecessary state changes when navigating away from Explore.
```

File: public/app/features/explore/ExplorePage.test.tsx
Comment (ifrost) [inline public/app/features/explore/ExplorePage.test.tsx:136]: Could you create a task in the backlog for it and cross reference it here?

File: public/app/features/explore/ExplorePaneContainer.tsx
Comment (ifrost) [inline public/app/features/explore/ExplorePaneContainer.tsx:33]: Thanks for writing it down 👍  It'd be worth to have a GitHub Issue to look deeper into this. Also, we usually use  `/**` comment blocks with `*` on each line.

File: public/app/features/explore/spec/helper/setup.tsx
Comment (ifrost) [inline public/app/features/explore/spec/helper/setup.tsx:235]: Can we remove `multi` param now? (Also from all usages)

File: public/app/features/explore/state/explorePane.ts
Comment (ifrost) [inline public/app/features/explore/state/explorePane.ts:143]: Do we still need this comment? Or maybe it should be moved to `useStateSync`?

File: public/app/features/explore/state/main.ts
Comment (ifrost) [inline public/app/features/explore/state/main.ts:70]: Should we have a task in the backlog for it?

File: public/app/features/explore/state/query.ts
Comment (ifrost) [inline public/app/features/explore/state/query.ts:323]: OMG 😨

Comment (ifrost) [reply public/app/features/explore/state/query.ts:323]: This will change the behavior but to something that actually makes sense 👍 I wonder if users got used to the old behavior that now they may be surprised 🤔 Maybe we could add it to the changelog?

File: public/app/features/explore/state/utils.test.ts
Comment (ifrost) [inline public/app/features/explore/state/utils.test.ts:67]: Assertions in both tests are the same, they could be extracted to a seperate function.

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync.ts:48]: Should we also check for keys in `panes` param?

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync.ts:165]:
```suggestion
    // Here we want to initialize each pane initial data, whether it comes
```

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync.ts:178]:
```suggestion
                // FIXME: In theory, given the Grafana datasource will always be present, this should always be defined.
```

Do we need `FIXME` here?

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync.ts:345]: Is it still relevant? It handles `panelState` changes, doesn't it?

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync.ts:32]: Do we need a task for adding unit tests?

**Review by ifrost — APPROVED** [review]
Wow 🤩  Love we get rid of all `saveState` and weird params passed to decide what/when is updated. It's way easier to see what's going on. Well done!

File: public/app/features/explore/ExplorePaneContainer.tsx
Comment (Elfo404) [reply public/app/features/explore/ExplorePaneContainer.tsx:33]: comments starting with`/**` are used to document functions with jsdoc (so that it shows up when doing autocomplete for instance). this is just a multiline comment explaining why that was needed. how would having a jsdoc block help here?

File: public/app/features/explore/ExplorePage.test.tsx
Comment (Elfo404) [reply public/app/features/explore/ExplorePage.test.tsx:136]: this is gonna be refactored anyway in the next PR that introduces migrations as each one of them will have to be tested against their specific url format.

File: public/app/features/explore/hooks/useStateSync.ts
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync.ts:32]: all the logic in here is already covered by unit tests in ExplorePage.test.tsx. with the migration PR i'll be extracting tests that are URL-specific to their own suite

File: public/app/features/explore/state/utils.test.ts
Comment (Elfo404) [reply public/app/features/explore/state/utils.test.ts:67]: i think separating assertions in tests from the actual test body it's a bad practice most of the times as it makes both debugging tests harder and errors harder to follow should the test itself stop working for some reason.

File: public/app/features/explore/hooks/useStateSync.ts
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync.ts]: good point. removed it in cc11aee646

File: public/app/features/explore/state/main.ts
Comment (Elfo404) [reply public/app/features/explore/state/main.ts:70]: will create tasks for each of the pending todo items once the PR is merged

File: public/app/features/explore/spec/helper/setup.tsx
Comment (Elfo404) [reply public/app/features/explore/spec/helper/setup.tsx:235]: good catch! updated it in 5e28f63909

File: public/app/features/explore/hooks/useStateSync.ts
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync.ts:48]: the `panes` params doesn't exist yet, it will be introduced in a follow up PR alongside the migrations.

File: public/app/features/explore/ExplorePaneContainer.tsx
Comment (ifrost) [reply public/app/features/explore/ExplorePaneContainer.tsx:33]: > how would having a jsdoc block help here?

It looks relevant to the component so works as a function doc to me. I'm fine with both, it's more about style convention - I think for multiline comments we use `//` on each line, and I haven't seen much `/* */` for multi-lines. It's more just a personal preference, as there's no linting rule to enforce anything.

File: public/app/features/explore/hooks/useStateSync.ts
Comment (ifrost) [reply public/app/features/explore/hooks/useStateSync.ts:48]: Coolio, just asking because there was already some code for handling pane keys.

Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync.ts:48]: the other `panes` key used in the PR is referencing the panes object in redux, this one is referencing the query params

File: packages/grafana-data/src/types/explore.ts
Comment (Elfo404) [reply packages/grafana-data/src/types/explore.ts]: nowhere, must be a leftover, good catch! fixed in 6be28b8e9d

File: public/app/features/explore/state/query.ts
Comment (Elfo404) [reply public/app/features/explore/state/query.ts:323]: opened https://github.com/grafana/grafana/pull/69643 for this

**Pull request merged**
